src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -676,6 +678,15 @@ void MeshForwarder::HandleMesh(FrameData &aFrameData, const Mac::Address &aMacSo
     meshAddrs.mDestination.SetShort(meshHeader.GetDestination());
 
     UpdateRoutes(aFrameData, meshAddrs);
+    macSource  = Get<RouterTable>().FindRouterByRloc16(aMacSource.GetShort());
+    meshSource = Get<RouterTable>().FindRouterByRloc16(meshAddrs.mSource.GetShort());
+    VerifyOrExit((macSource != nullptr && meshSource != nullptr), error = kErrorFailed);
```

Comment:
I suggest not using `VerifyOrExit()` here as we may unintentionally skip the rest of the code below (and fail to process a mesh header frame). We can check that macSource and meshSource are non-null before trying to update

One situation I was considering is if the mesh source is an FTD child and not a router. In that case, I expected that `FindRouterByRloc16()` would fail and return `nullptr`. 

However, upon reviewing the code, I noticed that `FindRouterByRloc16()` actually converts the given RLOC16 to a router ID. Thus, in this scenario, we would get the parent of the FTD child. This seems to be an unintended side-effect of the current implementation of `FindRouterByRloc16`, and therefore, I would not recommend assuming this behavior.

Reply:
For `VerifyOrExit()`, I think your suggestion is great and I will remove it.

For the second issue, here is my perspective:

Firstly, the unicast packets from a child are forwarded by its parent (router). From the actual packet captures, I found that for a packet sent from a child, the originator in the mesh header during forwarding is always its parent. Therefore, in `HandleMesh`, the RLOC16 returned by `meshAddrs.mSource.GetShort()` does not belong to the child, but to its parent.

Moreover, I think it is reasonable to get the parent when checking the router by a child's RLOC16. In the code, obtaining the router ID from the RLOC16 is done by shifting right 10 bits. For example, for a router with RLOC16 0xdc00 and its child with RLOC16 0xdc01, the result of the right shift by 10 bits will indeed be the same. Logically, since all unicast packets from a child are forwarded by its parent, the behavior of "***the router obtained through the child's RLOC16 is the child's parent***" is reasonable.

Do you have any other thoughts? What concerns do you have regarding the content of this pull request?

Reply:
I have found that when a FTD child sending a `Address Solicit` to become a router, a child's RCLO16 will apper in `HandleMesh` of the forwarding router. But I think it does not influence this pull request because meshSource always points a router rather than a child.

Reply:
> Do you think a child's RLOC16 would appear in HandleMesh?

Yes. I think a child's RLOC16 can appear in `HandleMesh` as `meshAddrs.mSource`. I think the same applies to `meshAddrs.mDestination`, although it may not be relevant in this specific context.


> the behavior of "the router obtained through the child's RLOC16 is the child's parent" is reasonable.

Yes. I agree. What we want here in case `meshAddr.mSource` is a child to use its parent when updating the next hop (route table).

> Do you have any other thoughts? What concerns do you have regarding the content of this pull request?

I don't like the use of `FindRouterByRloc16()` because it today happens to implicitly return the parent router when given a child's RLOC16. This behavior may change in the future (returning `null` for example), potentially breaking this code.

I suggest we be more explicit and determine the router ID directly from `meshAddr.mSource`. Then, we can use `FindRouterById()` instead. This approach clearly indicates that we're retrieving the router itself (if `mSource` is a router) or the parent (if `mSource` is a child).

Reply:
> Yes. I think a child's RLOC16 can appear in `HandleMesh` as `meshAddrs.mSource`. I think the same applies to `meshAddrs.mDestination`, although it may not be relevant in this specific context.
>
Yes, you are right. I apologize for the mistake: the device I initially used to test was an MTD child instead of an FTD child, which led me to mistakenly believe that a child's RLOC16 would not appear in HandleMesh.

I think we can pause this discussion because I discovered something else while looking through the code, which made me consider whether there might be a more comprehensive solution to replace this pull request. I will open a new conversation to discuss this discovery.